crates/ruff_python_formatter/src/verbatim.rs

```diff
@@ -573,33 +596,275 @@ impl Format<PyFormatContext<'_>> for TrailingFormatOffComment<'_> {
+                    // Ignore if inside an expression
+                    Tok::NonLogicalNewline if parens == 0 => break (range.start(), range.end()),
+                    Tok::NonLogicalNewline => {
+                        contains_newlines = ContainsNewlines::Yes;
+                    }
+                    Tok::Lbrace | Tok::Lpar | Tok::Lsqb => {
+                        parens = parens.saturating_add(1);
+                    }
+                    Tok::Rbrace | Tok::Rpar | Tok::Rsqb => {
+                        parens = parens.saturating_sub(1);
+                    }
+                    Tok::String { value, .. } if value.contains(['\n', '\r']) => {
+                        contains_newlines = ContainsNewlines::Yes;
+                    }
+                    _ => {}
+                },
+                None => {
+                    // Is there any remaining content after the last token? If so, return it as the last
+                    // logical line.
+                    return if self.last_line_end < self.content_end {
+                        let content_start = self.last_line_end;
+                        self.last_line_end = self.content_end;
+                        Some(Ok(LogicalLine {
+                            content_range: TextRange::new(content_start, self.content_end),
+                            contains_newlines,
+                            has_trailing_newline: false,
+                        }))
+                    } else {
+                        None
+                    };
+                }
+                Some(Err(_)) => {
+                    return Some(Err(FormatError::syntax_error(
+                        "Unexpected token when lexing verbatim statement range.",
+                    )))
+                }
+            }
+        };
+
+        let line_start = self.last_line_end;
+        self.last_line_end = full_end;
+
+        Some(Ok(LogicalLine {
+            content_range: TextRange::new(line_start, content_end),
+            contains_newlines,
+            has_trailing_newline: true,
+        }))
+    }
+}
+
+impl<I> FusedIterator for LogicalLinesIter<I> where I: Iterator<Item = LexResult> {}
+
+/// A logical line or a comment (or form feed only) line
+struct LogicalLine {
+    /// The range of this lines content (excluding the trailing newline)
+    content_range: TextRange,
+    /// Whether the content in `content_range` contains any newlines.
+    contains_newlines: ContainsNewlines,
+    /// Does this logical line have a trailing newline or does it just happen to be the last line.
+    has_trailing_newline: bool,
+}
+
+impl Ranged for LogicalLine {
+    fn range(&self) -> TextRange {
+        self.content_range
+    }
+}
+
+struct VerbatimText {
+    verbatim_range: TextRange,
+    contains_newlines: ContainsNewlines,
+}
+
+fn verbatim_text<T>(item: T, contains_newlines: ContainsNewlines) -> VerbatimText
 where
     T: Ranged,
```

Comment:
Uhm, not consciously. I generally avoid using `impl` (but there are exceptions), and lean towards `where` because I find them more readable when more complicated constraints become necessary.